text/0000-toolchain.md

```diff
@@ -0,0 +1,193 @@
+
+### User tools: `surge`
+
+An end user of the [surge.sh](https://surge.sh) service installs their CLI tool that is deployed as an npm package:
+```
+notion install surge
+```
+In this user story, the `surge` tool is published with `"engines": "11"` in its manifest, and at the time the command is run, the latest 11.x version of Node is 11.4.0. The CLI tool is installed in the user toolchain with Node 11.4.0 set as its default engine.
+
+From this point on, unless the user changes their toolchain, running
+```
+surge
+```
+from the command-line always runs this tool using Node 11.4.0.
+
+### User tools in projects: `pexx`
+
+The [project-explorer](https://sdras.github.io/project-explorer-site/) tool is a useful CLI tool that lets you inspect JS projects, but is not itself a tool that a project would use in its scripts.
+
+A user can install this tool to their toolchain:
+```
+notion install project-explorer
+```
+In this user story, the `project-explorer` manifest is published with `"engines": "8"` in its manifest. At the time the tool is installed, the latest 8.x version of Node is 8.15.0. So the `pexx` tool is installed with Node 8.15.0 as its default engine.
+
+When the user runs
+```
+pexx myproject
+```
+the `pexx` binary runs with Node 8.15.0—even if `myproject` specifies a different version of Node in its `"toolchain"` spec.
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+This section lists the set of concepts that users may encounter using Notion. The first two, **tools** and the **toolchain**, are central to using Notion. The latter two, **shims** and **engines**, are lower-level primitives that may be helpful for more implementation-oriented users.
+
+## Tools
+
+There are three types of tools:
+
+- **Node runtime:** The version of Node itself, which in particular dictates which `node` binary gets invoked.
+- **Package manager:** A version of npm or a version of Yarn. (In the future we may want to add support for other package managers such as pnpm or tink.)
+- **Package binary:** An executable published and distributed as part of an npm package.
+
+## Toolchain
+
+The toolchain is a set of tools the user has installed for use at the command-line console. The user adds to their toolchain with `notion install` and removes with `notion uninstall`.
+
+Each tool installed in the toolchain has a default version and engine, which can be overridden when the tool is invoked in a project that has a dependency on that tool.
+
+## Primitive: Shims
+
+Notion shims intercept calls to tools and redirect execution to the right executable based on the environment and current directory.
+
+Users may already be familiar with shims, since these are commonly used by other version managers.
+
+## Primitive: Engines
+
+An **engine spec** is a complete description of a version of the Node platform:
+
+- An exact version of the Node runtime.
+- An optional exact version of npm.
```

Comment:
One point of clarification I'd love: is this a "both are always optional" or is it "always at least one of these"? An engine without one of both would seem to be broken.

Put in concrete Rust terms, is it A or B here:

```rust
// A
struct Engine {
    node_version: String,
    npm_version: Option<String>,
    yarn_version: Option<String>,
}

// B
struct Engine {
    node_version: String,
    package_manager: PackageManager
}

enum PackageManager {
    Npm { version: String },
    Yarn { version: String },
}
```

Reply:
It's currently A (In fact, see https://github.com/notion-cli/notion/blob/master/crates/notion-core/src/platform/mod.rs#L14 which is almost exactly that), because Node always comes bundled with a version of `npm`. So if neither `npm` nor `yarn` is specified, we default to the version of `npm` that is bundled with Node.